### PR TITLE
Add WebAssembly specs

### PR DESCRIPTION
**Pull request by lbguilherme**
This PR depends on #12569 and adds a few more wasm32 patches to get specs running.

62% of the stdlib specs are able to run.

Note that `expect_raises` is disabled since WebAssembly doesn't have exceptions yet. So some of those are incomplete.

I'm opening as a WIP to get a CI run.

### CODE REVIEW

**Comment by lbguilherme** [comment]
It's greeen!

Ready for review.

File: .github/workflows/wasm32.yml
Comment (HertzDevil) [inline .github/workflows/wasm32.yml:30]: The download checksum here should be verified

File: .github/workflows/wasm32.yml
Comment (straight-shoota) [inline .github/workflows/wasm32.yml:21]: Can we pin this to a specific version?
```suggestion
          curl https://wasmtime.dev/install.sh -sSf | bash -s - --version v1.0.1
```

**Comment by paulocoghi** [comment]
Anything holding #12569 and, consequently, this PR?

**Comment by straight-shoota** [comment]
@paulocoghi https://github.com/crystal-lang/crystal/pull/12569 is in the merge queue for 1.7.0 which we'll start to process now that 1.6.1 is released.
This PR is waiting for another approval from a Core Team member.

File: src/spec/expectations.cr
Comment (beta-ziliani) [inline src/spec/expectations.cr:385]: I don't understand this change. Why would you want to have the original expect_raises in wasm32?

Comment (lbguilherme) [reply src/spec/expectations.cr:385]: Otherwise the smoke tests fail to compile. This mock changes the return type (from the exception instance to nil) and a few of the spec files fail to compile. And if I don't mock, pretty much all tests fail to run.

Comment (beta-ziliani) [reply src/spec/expectations.cr:385]: Ah, actually the real question is what's the need for smoke tests if this PR adds a proper test

Comment (lbguilherme) [reply src/spec/expectations.cr:385]: Ah, good question... I guess because it compiles all tests instead of ~half of them? It will get less and less useful as time goes by. 

Shall I remove the smoke tests?

Comment (straight-shoota) [reply src/spec/expectations.cr:385]: Yes, when adding *actually* running test we can remove the compile-only smoke tests. Even when they only have partial coverage. It's certainly better than anything a smoke test could provide.

Comment (straight-shoota) [reply src/spec/expectations.cr:385]: (Arguably, there's still a low chance to miss changes affecting currently untested code, which a smoke test would catch. But more and more running specs will be enabled in the future at which point such defects will be detected and fixed.)

Comment (beta-ziliani) [reply src/spec/expectations.cr]: Minor comment: I'd say the chances are not that low. Yet, in a vast majority the error will be fixed raising a "not implemented", which isn't a big difference. So yeah, test > smoke.

Comment (straight-shoota) [reply src/spec/expectations.cr]: Yeah, I meant a low chance to miss a relevant change. One that accidentally breaks something that used to work before and would be discovered in the smoke test.

File: .github/workflows/wasm32.yml
Comment (straight-shoota) [inline .github/workflows/wasm32.yml:21]: Would it make sense to use an install action?
I could only find https://github.com/marketplace/actions/set-up-wasmtime though, which doesn't seem to be maintained (at least there's been no update for 2 years).

**Review by straight-shoota — APPROVED** [review]

**Comment by lbguilherme** [comment]
`crystallang/crystal:1.5.0-build` contains `wasm-ld-10` and Crystal is built against LLVM `10.0.0`:

```
Crystal 1.5.0 [994c70b10] (2022-07-06)

LLVM: 10.0.0
Default target: x86_64-unknown-linux-gnu
```

`crystallang/crystal:1.6.1-build` contains `wasm-ld-10` and Crystal is built against LLVM `13.0.1`:

```
Crystal 1.6.1 [fc61bd678] (2022-10-21)

LLVM: 13.0.1
Default target: x86_64-unknown-linux-gnu
```

Unfortunately due to changes in the object format in LLVM an object produced by LLVM 13 can't be linked by the linker included in LLVM 10.

I'm now installing LLVM 13 there to ensure the correct linker is available.

**Comment by straight-shoota** [comment]
Oh, it seems we need to update LLVM in the build images.

File: .github/workflows/wasm32.yml
Comment (lbguilherme) [reply .github/workflows/wasm32.yml]: I wasn't updated recently but seems to work well. Added!

**Comment by straight-shoota** [comment]
I think we should probably merge this right away and refactor later once we have updated LLVM in the build images (ref https://github.com/crystal-lang/distribution-scripts/issues/206).

**Review by beta-ziliani — APPROVED** [review]
Great work, thanks Guilherme!

**Pull request merged**
